ngmix/prepsfmom/gaussmom.py

```diff
@@ -0,0 +1,352 @@
+    x = x.astype(np.float64)
+    y = y.astype(np.float64)
+    y -= row0
+    x -= col0
+    v = dvdrow*y + dvdcol*x
+    u = dudrow*y + dudcol*x
+
+    # now rescale by weight map, keeping the maximum value the same
+    rkf *= wgt
+    rkf *= (mval/rkf[mind])
+
+    # now build the moment kernels and their FFTs
+    rkxx = rkf * u**2
+    rkxy = rkf * u * v
+    rkyy = rkf * v**2
+
+    fkf = np.fft.fftn(rkf)
+    fkxx = np.fft.fftn(rkxx)
+    fkxy = np.fft.fftn(rkxy)
+    fkyy = np.fft.fftn(rkyy)
+
+    # the linear combinations here measure the moments proportional to the size
+    # and shears
+    return dict(
+        rkf=rkf,
+        rkr=rkxx + rkyy,
+        rkp=rkxx - rkyy,
+        rkc=2.0 * rkxy,
+        fkf=fkf,
+        fkr=fkxx + fkyy,
+        fkp=fkxx - fkyy,
+        fkc=2.0 * fkxy,
+    )
+
+
+def _measure_moments_fft(
+    im, inv_wgt, eff_pad_factor,
+    cen_row, cen_col,
+    kernels,
+    max_psf_frac,
+    psf_im,
+    psf_row_offset,
+    psf_col_offset,
+):
+    flags = 0
+    flagstr = ''
+
+    imfft = np.fft.fftn(im)
+
+    # we need to shift the FFT so that x = 0 is the center of the profile
+    # this is a phase shift in fourier space
+    # we have to do it for the profile and the kernel (and the PSF if needed)
+    f = np.fft.fftfreq(im.shape[0])
+    # this reshaping makes sure the arrays broadcast nicely
+    fx = f.reshape(1, -1)
+    fy = f.reshape(-1, 1)
+    kcen = 2.0 * np.pi * (fy*cen_row + fx*cen_col)
+    cen_phase = np.cos(kcen) + 1j*np.sin(kcen)
+    # instead of adjusting the kernels, we will apply the shift twice to the image
+    imfft *= cen_phase
+    imfft *= cen_phase
+
+    # first we shift the PSF to the object center
+    psf_imfft = np.fft.fftn(psf_im)
+    fx = f.reshape(1, -1)
+    fy = f.reshape(-1, 1)
+    kcen = 2.0 * np.pi * (fy*psf_row_offset + fx*psf_col_offset)
+    psf_cen_phase = np.cos(kcen) + 1j*np.sin(kcen)
+    psf_imfft *= psf_cen_phase
+
+    # now we apply the shift as above
+    psf_imfft *= cen_phase
+
+    # now we remove the PSF
+    # we truncate models below max_psf_frac of the maximum of the PSF profile
+    # set the PSF to 1 there to make sure we don't divide by zero
+    # set the kernels to zero there to ensure we do not use those modes
+    abs_psfimfft = np.abs(psf_imfft)
+    psf_zero_msk = abs_psfimfft <= max_psf_frac * np.max(abs_psfimfft)
+    if np.any(psf_zero_msk):
+        psf_imfft[psf_zero_msk] = 1.0
+        for k in kernels:
+            if k.startswith("f"):
+                kernels[k][psf_zero_msk] = 0.0
+
+    # deconvolve!
+    imfft /= psf_imfft
+
+    # finally we build the kernels, moments and their errors
+    df = f[1] - f[0]  # this is the area factor for the integral we are
+    # doing in Fourier space
+
+    # build the flux, radial, plus and cross kernels / moments
+    fkf = kernels["fkf"]
+    fkr = kernels["fkr"]
+    fkp = kernels["fkp"]
+    fkc = kernels["fkc"]
+    mf = np.sum(imfft * fkf).real * df**2
```

Comment:
I do apply a phase shift in fourier space to translate the kernels to the right spots for the sum. 

I don't see how this would cause extra noise though.

Reply:
OK

Reply:
I should have said, I don't understand how this would cause extra noise. I didn't mean to imply it doesn't. I don't know either way.